File: 0129-sum-root-to-leaf-numbers/0129-sum-root-to-leaf-numbers.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        st = []
        numstack = []
        num = 0
        s = 0
        
        while root or st:
            while root:  
                num = root.val + (num * 10)
                numstack.append(num)
                st.append(root)
                root = root.left       
            
            root = st.pop()
            num = numstack.pop()
            
            if root.right is None and root.left is None:
                s += num
            root = root.right
        
        return s

File: 0129-sum-root-to-leaf-numbers/0129-sum-root-to-leaf-numbers.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    int result;
    public int sumNumbers(TreeNode root) {
        if(root == null){
            return 0;
        }

        return inorder(root, 0);
        // return result;
    }

    private int inorder(TreeNode root, Integer rootSum){
        //base
        if(root == null){
            return 0;
        }
        
        if(root.left == null && root.right == null){
            return  (rootSum * 10) + root.val;
        }

        //logic
        return inorder(root.left, rootSum * 10 + root.val) + inorder(root.right, rootSum * 10 + root.val);
    }
}